### ISSUE TITLE
Call class and method and pass parameters using global route

### ISSUE THREAD

**Issue by iamdriz** [dannyvankooten/AltoRouter#144]
Is it possible to create a global route that will call a class and method and pass any optional parameters (including query strings)?

So for example let's say I have the following class acting as a controller and containing methods for actions (like in an MVC pattern). I also want to catch if the action isn't passed and default to the Index method in the controller.

~~~
class App
{
	public static function Index()
	{
		echo 'Index!';
		if(isset($_REQUEST['page'])) echo 'Page: ' . $_REQUEST['page'];
	}
	
	public static function Create()
	{
		echo 'Create!';
	}
	
	public static function Edit($id)
	{
		echo 'Edit! '.$id;
	}
	
	public static function Delete($id)
	{
		echo 'Delete! '.$id;
	}
	
	public static function View($id)
	{
		echo 'View! '.$id;
	}
}
~~~

Then I want the following URLs to all be catch able and hook up to the above class methods:

~~~
localhost/Demo/LearnBuddy/v1/App/
localhost/Demo/LearnBuddy/v1/App/?page=2
localhost/Demo/LearnBuddy/v1/App/Index/
localhost/Demo/LearnBuddy/v1/App/Index/?page=2
localhost/Demo/LearnBuddy/v1/App/Create/
localhost/Demo/LearnBuddy/v1/App/Edit/22/
localhost/Demo/LearnBuddy/v1/App/Delete/22/
localhost/Demo/LearnBuddy/v1/App/View/22/
localhost/Demo/LearnBuddy/v1/App/View/22/
~~~

Using something along the lines of:

~~~
$router = new AltoRouter();

// home
$router->map( 'GET', '/', function() {	
	echo 'home';
});

// call the index if no action is passed
$router->map( 'GET', '/[a:controller]/', function( $controller ) {	
	if( method_exists( $controller, $action ) )
		$controller::Index();
	else
		echo 'missing';
});

// call controller and action and pass any parameters
$router->map( 'GET', '/[a:controller]/[a:action]/[a:params]', function( $controller, $action, $params ) {
	if( method_exists( $controller, $action ) )
		$controller::$action($params);
	else
		echo 'missing';
});

$match = $router->match();

var_dump($match);

// call closure or throw error
if( $match && is_callable( $match['target'] ) )
{
	call_user_func_array( $match['target'], $match['params'] ); 
}
else
{
	echo 'error';
}
~~~

So in short, how can I amend the above code to catch controller and action with optional params and query strings? Currently I'm just getting `bool(false)` when I `var_dump($match);`.

This is on a Windows machine running IIS (the web config seems to fine, as all URLs are hitting the index.php file so I don't think that is causing none of them to work.).

**Comment by irecinius** [comment]
BasePath is wrong, thats why its not hitting any match

**Comment by iamdriz** [comment]
What should the base path be?

**Comment by iamdriz** [comment]
@irecinius basePath shouldn't of been in the code example, I still get the same error.

**Comment by koenpunt** [comment]
I tested your code, and I'm quite sure it has to do with the trailing slash. One of your routes is defined with, and one without, and AltoRouter is quite strict on that. So please make sure you request the URL with a trailing slash OR remove the trailing slash from the route definition.

**Comment by koenpunt** [comment]
Also, matching with and without action, can be a single route by making the action optional by appending a questionmark. For example:

```php
$router->map('GET', '/[a:controller]/[a:action]?', function($controller, $action = null) {
  $action = $action !== null ? $action : 'Index';
  if(method_exists($controller, $action)) {
    $controller::$action();
  } else {
    echo 'missing';
  }
});
```
